### ISSUE TITLE
100% sparsity (0 NNZ) in WRPN

### ISSUE THREAD

**Issue by KangkyuPark** [NervanaSystems/distiller#308]
First of all, thanks for providing distiller. This is very helpful for me:)

I met a problem(?) when training WRPN.
I trained resnet-18 and resnet-50 using WRPN quantizer.
I expected there are sparse weights in WRPN whereas there are no sparse weights in DoReFa-Net and PACT.
The problem is the sparsity in WRPN is very high.
In ResNet-18, the sparsity is 75~98% in a layer; even the 1st layer of the 1st module (module.layer1.0.conv1.weight) has 80% sparsity.
Furthermore, In ResNet-50 with WRPN, there are layers which have 100% sparsity..
I trained the two networks with 4-bit precision for both activation and weight.
The command line is
 "Command line: compress_classifier.py -a preact_resnet18 /home/tools/ImageNet/Data2/ -j 16 -p 10 --epochs 200 --compress=../quantization/quant_aware_train/preact_resnet18_imagenet_wrpn.yaml --gpus=0 --batch-size 128 --out-dir=logs/wrpn_resnet18_w4a4".

Are the networks trained properly?
If not, how much the sparsity with WRPN when I trained the networks properly?

![resnet18](https://user-images.githubusercontent.com/24670054/60727129-2197d000-9f78-11e9-88e8-8d6fa1a9ac46.PNG)
![resnet50](https://user-images.githubusercontent.com/24670054/60727130-2197d000-9f78-11e9-8243-3e0af82586dd.PNG)
![yaml](https://user-images.githubusercontent.com/24670054/60727131-2197d000-9f78-11e9-90f2-1e031b6f1219.PNG)

**Comment by levzlotnik** [comment]
Hi @KangkyuPark ,

Sorry for the late reply - 
The sparsity of `**.weight` is so high because it's quantized, and the quantized `int[<8]` may have a greater chance of being 0.
Since the summary isn't aware of the `*.weight` being the quantized version of `*.float_weight`, it includes it into the calculation.

Thanks,
Lev

**Comment by guyjacob** [comment]
@KangkyuPark, reading your post again, I think @levzlotnik and I might have mis-understood your original question.

First, to answer your question of "are networks are trained properly?" - I don't see an issue with your command line or scheduler.

**Now, to the issue of high rates of zeros:**

First, I should note that the `*.weight` parameter is actually not the quantized version of `*.float_weight`, per se. It is the result of taking `*.float_weight` parameter and performing quantization + de-quantization. So the `*.weight` parameter does not contain integer values, as may have been implied by @levzlotnik's answer.

Now, as you said, it is not surprising to see zero weights when using WRPN as opposed to DoReFa and PACT. That's because WRPN does simple symmetric quantization of the weights - so if an element became zero after quantization, it remains zero after de-quantization as well. On the other hand, the transformation employed by DoReFa and PACT (it's the same in both) actually prevents the weights from becoming zero after de-quantization.

Also, note that WRPN quantizes the weights assuming a floating-point range between [-1 .. 1]. If the weights are actually in a much narrower range, you'd get a very coarse quantization which could indeed have a very high rate of 0's (up to the entire layer). Seeing as ResNet weights are initialized from a normal distribution with mean 0 and std of sqrt(2 / fan_out) - it actually might not be so surprising we're ending up with lots of zeros.

To summarize - I'd say that the implementation of WRPN in Distiller faithfully follows the method described in the paper. As discussed above, there could be issues with the method itself, which could result in very high rates of 0's. So I don't find what you're seeing surprising.
On one hand, one could consider the high level of sparsity a "feature". But, if the quantization is too coarse, that could obviously limit what the quantized model is able to represent and have a negative impact on overall accuracy. It could be interesting to try quantizing using the actual range of the weights instead of just assuming [-1 .. 1]. But that wouldn't be WRPN anymore :)

**Comment by guyjacob** [comment]
Closing this. @KangkyuPark please re-open if you still have questions on this.
